File: ios/App/URB/Commands/Network/NetworkStatusCommand.swift
import Foundation
import Network

final class UrbNetworkStatusMonitor {
    private let monitor = NWPathMonitor()
    private let queue = DispatchQueue(label: "urb.network.monitor")
    private let eventEmitter: UrbEventEmitter
    private var started = false

    init(eventEmitter: UrbEventEmitter) {
        self.eventEmitter = eventEmitter
    }

    func startIfNeeded() {
        guard !started else { return }
        started = true
        monitor.pathUpdateHandler = { [weak self] path in
            guard let self else { return }
            self.eventEmitter.emit(name: "network:statusChange", payload: self.status(path: path))
        }
        monitor.start(queue: queue)
    }

    func current() -> [String: Any] {
        status(path: monitor.currentPath)
    }

    func cancel() {
        monitor.cancel()
    }

    private func status(path: NWPath) -> [String: Any] {
        let connected = path.status == .satisfied
        let type: String
        if !connected { type = "none" }
        else if path.usesInterfaceType(.wifi) { type = "wifi" }
        else if path.usesInterfaceType(.cellular) { type = "cellular" }
        else if path.usesInterfaceType(.wiredEthernet) { type = "ethernet" }
        else if path.usesInterfaceType(.other) { type = "vpn" }
        else { type = "unknown" }
        return ["connected": connected, "type": type, "expensive": path.isExpensive]
    }
}

final class UrbNetworkStatusCommand: UrbCommand {
    let name = "network:getStatus"
    private let monitor: UrbNetworkStatusMonitor
    init(monitor: UrbNetworkStatusMonitor) {
        self.monitor = monitor
        self.monitor.startIfNeeded()
    }
    func handle(request: UrbRequest, responder: UrbResponder) {
        responder.success(monitor.current())
    }
}

File: android/app/src/main/java/com/example/mywebview/urb/commands/network/NetworkStatusCommand.kt
package com.example.mywebview.urb.commands.network

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import com.example.mywebview.urb.core.UrbCommand
import com.example.mywebview.urb.bridge.UrbEventEmitter
import com.example.mywebview.urb.core.UrbRequest
import com.example.mywebview.urb.core.UrbResponder
import org.json.JSONObject

class NetworkStatusCommand(
    context: Context,
    private val eventEmitter: UrbEventEmitter,
) : UrbCommand {
    override val name = "network:getStatus"
    override val expectsResponse = true
    private val connectivityManager =
        context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
    private var registered = false
    private val callback = object : ConnectivityManager.NetworkCallback() {
        override fun onAvailable(network: Network) {
            emitStatus()
        }

        override fun onLost(network: Network) {
            emitStatus()
        }

        override fun onCapabilitiesChanged(
            network: Network,
            networkCapabilities: NetworkCapabilities,
        ) {
            emitStatus()
        }
    }

    fun register() {
        if (registered) return
        connectivityManager.registerDefaultNetworkCallback(callback)
        registered = true
        emitStatus()
    }

    fun unregister() {
        if (!registered) return
        connectivityManager.unregisterNetworkCallback(callback)
        registered = false
    }

    override fun handle(request: UrbRequest, responder: UrbResponder?) {
        responder?.success(currentStatus())
    }

    private fun emitStatus() {
        eventEmitter.emit(
            JSONObject()
                .put("channel", "urb")
                .put("name", "network:statusChange")
                .put("payload", currentStatus()),
        )
    }

    private fun currentStatus(): JSONObject {
        val network = connectivityManager.activeNetwork
        val capabilities = connectivityManager.getNetworkCapabilities(network)
        val connected = capabilities?.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) == true
        val type = when {
            !connected -> "none"
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> "wifi"
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> "cellular"
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) -> "ethernet"
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_VPN) -> "vpn"
            else -> "unknown"
        }
        val expensive = capabilities
            ?.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_METERED)
            ?.not()
            ?: false

        return JSONObject()
            .put("connected", connected)
            .put("type", type)
            .put("expensive", expensive)
    }
}
